File: royale/source/core/src/ExtendedData.cpp
/****************************************************************************\
 * Copyright (C) 2015 Infineon Technologies
 *
 * THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
 * KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
 * PARTICULAR PURPOSE.
 *
 \****************************************************************************/

#include <processing/ExtendedData.hpp>

using namespace royale;
using namespace royale::processing;

ExtendedData::ExtendedData() :
    m_depthData (nullptr),
    m_rawData (nullptr),
    m_intermediateData (nullptr)
{
}

ExtendedData::~ExtendedData()
{
}

void ExtendedData::setDepthData (const DepthData *data)
{
    m_depthData = data;
}

void ExtendedData::setRawData (const RawData *data)
{
    m_rawData = data;
}

void ExtendedData::setIntermediateData (const IntermediateData *data)
{
    m_intermediateData = data;
}

bool ExtendedData::hasDepthData() const
{
    return m_depthData != nullptr;
}

bool ExtendedData::hasRawData() const
{
    return m_rawData != nullptr;
}

bool ExtendedData::hasIntermediateData() const
{
    return m_intermediateData != nullptr;
}

const RawData *ExtendedData::getRawData() const
{
    return m_rawData;
}

const DepthData *ExtendedData::getDepthData() const
{
    return m_depthData;
}

const IntermediateData *ExtendedData::getIntermediateData() const
{
    return m_intermediateData;
}

File: royale/source/RoyaleDotNet/ExtendedData.cs
/****************************************************************************\
* Copyright (C) 2015 Infineon Technologies
*
* THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
* KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
* IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
* PARTICULAR PURPOSE.
*
\****************************************************************************/

using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace RoyaleDotNet
{
    /// <summary>
    /// Class for getting additional data to the standard depth data. The
    /// retrieval of this data requires L2 access. Please be aware that not
    /// all data is filled. Therefore, use the has* calls to check if data is provided.
    /// </summary>
    public sealed class ExtendedData
    {
        [StructLayout (LayoutKind.Sequential)]
        private struct NativeExtendedData
        {
            [MarshalAs (UnmanagedType.I1)] public bool hasDepthData;
            [MarshalAs (UnmanagedType.I1)] public bool hasIntermediateData;
            [MarshalAs (UnmanagedType.I1)] public bool hasRawData;
            public IntPtr depthData;
            public IntPtr intermediateData;
            public IntPtr rawData;
        }

        private DepthData m_depthData;
        private IntermediateData m_intermediateData;
        private RawData m_rawData;
        private bool m_hasDepthData;
        private bool m_hasIntermediateData;
        private bool m_hasRawData;

        internal ExtendedData (IntPtr data)
        {
            NativeExtendedData nativeExtendedData = (NativeExtendedData) Marshal.PtrToStructure (
                    data, typeof (NativeExtendedData));

            m_hasDepthData = nativeExtendedData.hasDepthData;
            m_hasIntermediateData = nativeExtendedData.hasIntermediateData;
            m_hasRawData = nativeExtendedData.hasRawData;

            if (m_hasDepthData)
            {
                m_depthData = DepthDataConverter.FromNativeData (nativeExtendedData.depthData);
            }

            if (m_hasIntermediateData)
            {
                m_intermediateData = IntermediateDataConverter.FromNativeData (nativeExtendedData.intermediateData);
            }

            if (m_hasRawData)
            {
                m_rawData = RawDataConverter.FromNativeData (nativeExtendedData.rawData);
            }
        }

        public DepthData DepthData
        {
            get
            {
                return m_depthData;
            }
        }

        public IntermediateData IntermediateData
        {
            get
            {
                return m_intermediateData;
            }
        }

        public RawData RawData
        {
            get
            {
                return m_rawData;
            }
        }

        public bool HasDepthData
        {
            get
            {
                return m_hasDepthData;
            }
        }

        public bool HasIntermediateData
        {
            get
            {
                return m_hasIntermediateData;
            }
        }

        public bool HasRawData
        {
            get
            {
                return m_hasRawData;
            }
        }
    }
}
